### PR TITLE
Improve templates

### PR DESCRIPTION
**Pull request by okainov**
Improved templates a bit:
- Use better quality PNG for `png1` template
- Put all the text for `png1` template dynamically, fix wrong RU translations
- Added proper UTF wordwrap function
- Add `png5` template based on existing picture from `png1` folder

I checked it locally with bare PHP, I assume it should be fine with production env as well...

Samples:

`png5`: 
image

`png1`:
image

------------

Please do not squash, I expect my commits to be atomic and focused enough to stay as is :)

### CODE REVIEW

**Comment by okainov** [comment]
The only thing I'm really not sure is where to put new function... I created website/templates/labels/common.php, but probably there should be better place for it...

**Comment by kumy** [comment]
@boly38 may you please review this PR thanks? ;)

**Comment by okainov** [comment]
I fixed all formatting issues except this one. It doesn't make any sense for me as it reduces code readability.

```
-                        $line .= mb_substr($actual, 0, $width) . $break;
+                        $line .= mb_substr($actual, 0, $width).$break;
```

**Comment by boly38** [comment]
Oh cool a contrib on template :)
If you'll wait from me, then I will try this week to see that PR

**Comment by okainov** [comment]
@boly38 so far it's only existing templates which were not exposed to public for some reason :) However I do hope to get a template from our active geocacher so I may add some completely new ones later (depends how will it go)

**Comment by okainov** [comment]
Added two more templates from SanSanchoz

horizontal `png6`:
image

vertical `png7`:
image

File: website/templates/labels/png1/conf.php [hunk 1]
```diff
@@ -1,20 +1,38 @@
 <?php
 
-$imgname = "$kret_szablon/label.jpg";
-$img = imagecreatefromjpeg($imgname); /* Attempt to open */
+require_once '../common.php';
```
Comment (boly38) [inline website/templates/labels/png1/conf.php]: to have functional label on my side (local test with docker + gkShell), I had to put
```
require_once 'common.php';`
```

File: website/templates/labels/png1/conf.php [hunk 2]
```diff
@@ -1,20 +1,38 @@
 <?php
```
Comment (boly38) [inline website/templates/labels/png1/conf.php]: (minor) to help debug, add a comment with template name ?

````
// Modern :: Wallson (beta)
````

File: website/templates/labels/png1/conf.php [hunk 3]
```diff
@@ -1,20 +1,38 @@
...
-imagettftext($img, 15, 0, 252, 122, $czarny, $font, $kret_owner);
-imagettftext($img, 10, 0, 252, 212, $czarny, $font, wordwrap(stripcslashes(strip_tags($kret_opis, '<img>')), 30));
+imagettftext($img, 35, 0, 600, 100, $czarny, $font, "$kret_nazwa");
+imagettftext($img, 35, 0, 600, 287, $czarny, $font, $kret_owner);
+imagettftext($img, 25, 0, 600, 500, $czarny, $font, mb_wordwrap(stripcslashes(strip_tags($kret_opis, '<img>')), 35));
+
+imagettftext($img, 35, 0, 1420, 100, $czarny, $font, $kret_tracking);
+imagettftext($img, 35, 0, 2010, 100, $czarny, $font, $kret_id);
+
+$manual_strings = array(
```
Comment (boly38) [inline website/templates/labels/png1/conf.php]: @kumy did you know if [gettext](https://www.php.net/manual/fr/function.gettext.php) could enforce locale to use for a given translation ? 

thus in order to take advantage of the power of [crowdin](https://crowdin.com/project/geokrety)

ie having something like:

```` 
  'pl' => _($englishManualContent,'pl'),
  'en' => _($englishManualContent,'en'),
  'de' => _($englishManualContent,'de'),
  'cz' => _($englishManualContent,'cz'),
  'fr' => _($englishManualContent,'fr'),
  'ru' => _($englishManualContent,'ru')
```` 

(this feature seems not available, and switching locale is not a good way too, so Ignore this comment)

File: website/templates/labels/png1/conf.php [hunk 4]
```diff
@@ -1,20 +1,38 @@
...
+    'pl' => 'Instrukcja obsługi: 1. Zabierz geokreta. Zanotuj tracking code 2. Ukryj kreta w innym keszu. 3. Zarejestruj jego podróż na stronie https://geokrety.org',
+    'en' => "User's manual: 1. Take this GeoKret. Please note down his Tracking Code. 2. Hide in another cache. 3. Register the trip at https://geokrety.org",
+    'de' => 'Anleitung: 1. Nimm den GeoKret mit and notiere Dir den Tracking Code. 2. Verstecke ihn wieder in einem anderen Cache. 3. Logge seine Reise auf https://geokrety.org',
+    'cz' => 'Navod pro uživatele: 1. Vem Geokrtka. Poznamenej si jeho Tracking Code. 2. Schovej ho v jiné kešce. 3. Registruj cestu na https://geokrety.org',
+    'fr' => "Manuel de l'utilisateur: 1. Prenez cette GeoKret. Noteen bas de son code de suivi. 2. Masquer dans une autre cache. 3. S\'inscrire au voyage https://geokrety.org",
```
Comment (boly38) [inline website/templates/labels/png1/conf.php]: fr: 
````
Manuel de l'utilisateur: 1. Prenez ce GeoKret et notez son code de suivi. 2. Déposez le dans une autre cache. 3. Loguez son voyage sur https://geokrety.org

File: website/templates/labels/png5/conf.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+require_once '../common.php';
```
Comment (boly38) [inline website/templates/labels/png5/conf.php]:
```
require_once 'common.php';
````

File: website/templates/labels/png5/conf.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
+<?php
```
Comment (boly38) [inline website/templates/labels/png5/conf.php]: Beautiful one

I like the QRCode on both sides :)

add a comment 
// Modern :: Wallson 2 (with QR)

File: website/templates/labels/png6/conf.php [hunk 7]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+require_once '../common.php';
```
Comment (boly38) [inline website/templates/labels/png6/conf.php]:
```` 
require_once 'common.php';
````

File: website/templates/labels/png6/conf.php [hunk 8]
```diff
@@ -0,0 +1,28 @@
+<?php
```
Comment (boly38) [inline website/templates/labels/png6/conf.php]: :+1:

File: website/templates/labels/png1/conf.php (hunk 2 above)
Comment (okainov) [reply website/templates/labels/png1/conf.php]: done

File: website/templates/labels/png1/conf.php (hunk 1 above)
Comment (okainov) [reply website/templates/labels/png1/conf.php]: Sounds like black magic, but whatever, changed =)

File: website/templates/labels/png1/conf.php (hunk 4 above)
Comment (okainov) [reply website/templates/labels/png1/conf.php]: done

File: website/templates/labels/png5/conf.php (hunk 5 above)
Comment (okainov) [reply website/templates/labels/png5/conf.php]: done

File: website/templates/labels/png5/conf.php (hunk 6 above)
Comment (okainov) [reply website/templates/labels/png5/conf.php]: done

File: website/templates/labels/png6/conf.php (hunk 7 above)
Comment (okainov) [reply website/templates/labels/png6/conf.php]: done

**Comment by okainov** [comment]
But about this `common.php`  - I'm still pretty sure that there should be kind of file with common utility functions already - should I move my stuff this somewhere there?

File: website/templates/labels/png5/conf.php (hunk 6 above)
Comment (okainov) [reply website/templates/labels/png5/conf.php]: I also think it's really nice picture, wonder why this image was just laying around for quite some time without usage =)

File: website/templates/labels/png1/conf.php (hunk 3 above)
Comment (okainov) [reply website/templates/labels/png1/conf.php]: This piece of code for sure should be improved to use proper localization strings and not hardcoded text. But so far seems none of templates actually use this "language" selector on the page and also I haven't seen any localization examples around, so we can keep it for later IMO

**Comment by kumy** [comment]
> I fixed all formatting issues except this one. It doesn't make any sense for me as it reduces code readability.
> 
> ```
> -                        $line .= mb_substr($actual, 0, $width) . $break;
> +                        $line .= mb_substr($actual, 0, $width).$break;
> ```

If it breaks unit tests, it would be preferable to fix it, else just ignore it.

**Comment by boly38** [comment]
>But about this common.php - I'm still pretty sure that there should be kind of file with common utility functions already

sorry I didn't reply to this question
if you want a more deep refactoring, then
you could maybe add an utility class under website/lib directory (like ValidationUtils), and why not add related unit test(s) :)

In major cases of gkwebsite files you have a "require _sentry.php" at startup that includes (composer generated) autoload file. If you start this kind of update, then we could provide additional help if needed

**Comment by okainov** [comment]
>you could maybe add an utility class under website/lib directory (like ValidationService for example), and why not add related unit test(s) :)

done

**Comment by okainov** [comment]
Pushed some refactoring for #375 in the same PR, because I was basing it on current changes. Sorry for huge change -_-

File: website/labels.php [hunk 9]
```diff
@@ -33,7 +33,20 @@
...
-// ------ języki dostępne -----//
+// ------ available templates -----//
+
+$labelTemplates = '';
+
+//include all templates, needed to see all templates in declared_classes
+foreach (glob('lib/labels/*/*.php') as $filename) {
+    $className = pathinfo($filename)['filename'];
+    /** @var Template $tpl */
+    $tpl = new $className();
```
Comment (boly38) [inline website/labels.php]: to be a little bit more resilient, what about adding

````
    if (!class_exists($className)) {
      continue;
    }
    $tpl = new $className();
````

File: website/labels.php [hunk 10]
```diff
@@ -33,7 +33,20 @@
 $opis = strtr(strip_tags($opis), $conaco);
 $id = sprintf('GK%04X', $id);
 
-// ------ języki dostępne -----//
+// ------ available templates -----//
+
+$labelTemplates = '';
+
+//include all templates, needed to see all templates in declared_classes
+foreach (glob('lib/labels/*/*.php') as $filename) {
```
Comment (boly38) [inline website/labels.php]: :+1: this new mechanism of dynamic mapping is good improvement

File: website/lib/labels/classic/ClassicTemplate.php [hunk 11]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+class ClassicTemplate extends PngTemplate {
+    public function getName() {
+        return 'Classic :: filips';
```
Comment (boly38) [inline website/lib/labels/classic/ClassicTemplate.php]: this one render amost black on black on my side

image

(but this was already the case on prod)

File: website/lib/labels/Template.php [hunk 12]
```diff
@@ -0,0 +1,54 @@
+<?php
+
+abstract class Template {
```
Comment (boly38) [inline website/lib/labels/Template.php]: :1st_place_medal:

File: website/lib/labels/middleclassic/MiddleClassicTemplate.php [hunk 13]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+class MiddleClassicTemplate extends PngTemplate {
+    public function getName() {
+        return 'Middle classic :: filips';
```
Comment (boly38) [inline website/lib/labels/middleclassic/MiddleClassicTemplate.php]: same as previous (black on black on my side)

File: website/labels.php (hunk 10 above)
Comment (okainov) [reply website/labels.php]: Actually I was thinking that there should be some better mechanism for this ("get all subclasses of the class"), but I was not able to find anything better than this `glob` and hoping that filenames match classnames...

File: website/lib/labels/classic/ClassicTemplate.php (hunk 11 above)
Comment (okainov) [reply website/lib/labels/classic/ClassicTemplate.php]: I noticed, but I don't really know what to do with it. If you click "Print" in browser, that it's completely fine and good. So I consider the label itself as good enough.
If I disable some transparency flag during image generation, it starts to look ugly. Maybe I can play with source PNG, I can try... but shouldn't be a stopper anyway...

File: website/lib/labels/middleclassic/MiddleClassicTemplate.php (hunk 13 above)
Comment (okainov) [reply website/lib/labels/middleclassic/MiddleClassicTemplate.php]: same comment =)

File: website/lib/labels/classic/ClassicTemplate.php (hunk 11 above)
Comment (okainov) [reply website/lib/labels/classic/ClassicTemplate.php]: Fixed, added white background overlay

**Comment by okainov** [comment]
Now it's shown it browser correctly instead of transparent background

image

**Comment by okainov** [comment]
Any other comments? If everything is good, can it be merged and deployed?

File: website/lib/labels/classic/ClassicTemplate.php (hunk 11 above)
Comment (boly38) [reply website/lib/labels/classic/ClassicTemplate.php]: excellent! 
we feels less "buggy-like" behavior

File: website/labels.php (hunk 10 above)
Comment (boly38) [reply website/labels.php]: interesting question

following [this SO post](https://stackoverflow.com/questions/6671398/get-all-defined-classes-of-a-parent-class-in-php), I tried a little bit "instanceof" way without success, 
but maybe we could try to follow _registar pattern_ suggested by _Ahmad Hajjar_ ..

**Comment by boly38** [comment]
my side I ve very poor availability and I approve this PR (with or without last suggestion about "generic way to retrieve template") we could always produce a better code, but that's nice to deliver too ;)

File: website/labels.php (hunk 10 above)
Comment (boly38) [reply website/labels.php]: hum..not sure registar is a good way too .. I dont know

I finally think we would leave as you did

**Comment by kumy** [comment]
I also have little time this days. I will merge PR, it will be available at https://staging.geokrety.org 
We have some other PR to merge, then we'll deploy to prod after.

Thanks @okainov !

**Pull request merged**